replicatedhq/ship
# Add cert template functions

What I Did
------------
Added a set of template functions to allow creation and retrieval of RSA and ECDSA CAs and certificates.

These are:

`GetCaKey`:
Produces and returns the private key of a certificate authority. If a CA of this name has been generated before, it will not be regenerated.
 - Args
   - caName - the name this CA will be stored under
   - caType - the type of the key, defaulting to rsa-2048. Can be rsa-n (where n is size in bits), P256, P384 or P521.

`GetCaCert`:
Returns the cert of a generated certificate authority. If no CA of this name has been generated before, this function returns the empty string.
 - Args
   - caName - the name of the CA to retrieve

`GetKey`:
Produces and returns the private key of a cert. If a cert of this name has been generated before, it will not be regenerated.
- Args
   - certName - the name this cert will be stored under
   - caName - the name of the CA to use. If the CA does not exist, a new one will be created with this name and the same type as this certificate.
   - hosts - a comma delimited list of hosts this cert is valid for. An example is `abc.example.com,xyz.example.com`.
   - certType - the type of the key, defaulting to rsa-2048. Can be rsa-n (where n is size in bits), P256, P384 or P521.

`GetCert`:
Returns a generated cert. If no cert of this name has been generated before, this function returns the empty string.
- Args
   - certName - the name of the cert to retrieve


How I Did it
------------
Certs and CAs are generated with Cloudflare's [cfssl](https://github.com/cloudflare/cfssl) library. 

Generated Certs and CAs are stored by name within the ship state file, and future uses of the same name will yield the same cert.

How to verify it
------------
Unit tests to verify that the generated certs/CAs match a few desired properties (key matches cert, certs are signed by CA, certs cover requested hostnames, certs are not CAs, certs/CAs are unique)

Integration tests to verify that certs persist and are not regenerated when running `ship update` 

Description for the Changelog
------------
Added self-signed certificate generation template functions


Picture of a Boat (not required but encouraged)
------------

![USS Mahan (DD-364)](https://upload.wikimedia.org/wikipedia/commons/6/65/USSMahanDD364.jpg "USS Mahan (DD-364)")
